### DISCUSSION TITLE
MySQL & MariaDB do not work with QueryOrder enum values ending with `NULLS_LAST` or `NULLS_FIRST`

### DISCUSSION THREAD

**Discussion by ctfdavis** [mikro-orm/mikro-orm#4341] [General]
**Describe the bug**
When using the orderBy API on a MySQL or MariaDB database, if I provide any [`QueryOrder` enum](https://github.com/mikro-orm/mikro-orm/blob/master/packages/core/src/enums.ts#L55) values with the `NULLS_LAST` or `NULLS_FIRST` suffix, a SQL syntax error would result from the query:

```
…check the manual that corresponds to your MySQL server version for the right syntax to use near ‘nulls first' at line 1
```

The issue seems to be that MySQL and MariaDB do not natively support the `NULLS FIRST` and `NULLS LAST` options in the `ORDER BY` clause, leading to the syntax error mentioned above.

**To Reproduce**

1. clone [nestjs-realworld-example-app](https://github.com/mikro-orm/nestjs-realworld-example-app)
2. update mikro orm dependencies to 5.7.4
3. set up `config.ts`, `mikro-orm.config.ts` and a mysql database
4. change `QueryOrder.DESC` to `QueryOrder.DESC_NULLS_FIRST` on [line 57 in `article.service.ts`](https://github.com/mikro-orm/nestjs-realworld-example-app/blob/master/src/article/article.service.ts#L57)
5. start the app
6. fire a GET request to `/api/articles`

**Expected behavior**
- The QueryOrder enum values with the NULLS_LAST or NULLS_FIRST suffix should not throw an error when using MySQL and MariaDB.
- Support nulls first and nulls last sorting for MySQL and MariaDB.

**Versions**

- mikro-orm: 5.7.4
- @mikro-orm/mysql: 5.7.4

**Comment by B4nan** [comment]
`QueryOrder.DESC_NULLS_FIRST` and all the similar options in that enum are referring to an SQL standard, which is apparently not supported by MySQL/MariaDB. See this post for workarounds:

https://stackoverflow.com/questions/2051602/mysql-orderby-a-number-nulls-last

I think you can use any of them explicitly. Will take a closer look, if this can be supported automatically. The trick with `-` could be simple to implement, but as it works only on numeric columns, it's far away from perfect.

**Reply by ctfdavis** [reply]
Thanks for the reply! I think the second-most upvoted answer in the stack overflow question you referred to covers more column types. By the way, if I were to implement them explicitly, I am not sure where to start as there is not a method or option like `orderByRaw`?

**Reply by B4nan** [reply]
In v5 you can pass raw fragments there directly, it should be detected its a raw fragment, or you can try `em.raw()` for that. In v6, there is a new `raw` helper you can use (#4197).

**Reply by ctfdavis** [reply]
Thanks! I tried to pass a string to `orderBy` , but a property is automatically inserted before it, and it seems that I can’t change or remove it:

```ts
const qb = this.cRepository.qb(c);
qb .orderBy([‘ISNULL(c.title), c.title ASC' }]).getQuery();
// `c`.`id` is inserted right after 'order by’ 
// select `c`.* order by `c`.`id` isnull(c.title), c.title asc
```

As for `em.raw()` / `qb.raw()`, I am not sure about where to utilise it. It seems that `orderBy` does not work with the return type of `raw`?

**Reply by B4nan** [reply]
What you share won't even compile, as there is a syntax error, and looks invalid in general, order by clause needs to be either object or array of objects, not array of strings.

Try something like this:

```ts
qb.orderBy({ 'ISNULL(c.title)': 'ASC', 'c.title': 'ASC' });
```

> As for em.raw() / qb.raw(), I am not sure about where to utilise it. It seems that orderBy does not work with the return type of raw?

You could use `expr` to get around that, or just a type assertion.
